### ISSUE TITLE
how to combine apex.fp16_utils.fp16_optimizer with OSS

### ISSUE THREAD

**Issue by amsword** [facebookresearch/fairscale#740]
## ❓ Questions and Help
Is there some example or is it feasible to combine fp16_optimizer (https://nvidia.github.io/apex/_modules/apex/fp16_utils/fp16_optimizer.html#FP16_Optimizer) and OSS (from fairscale.optim.oss import OSS)

**Comment by anj-s** [comment]
I don't believe we have an example or test for this combination of APIs.  cc @blefaudeux 

We support torch.cuda.amp and have an [example](https://fairscale.readthedocs.io/en/latest/tutorials/oss.html) in our docs that show usage. It's also been [recommended](https://discuss.pytorch.org/t/torch-cuda-amp-vs-nvidia-apex/74994) that users use torch.cuda.amp for mixed precision training. 

Can you tell us a bit more about your use case and why apex APIs are needed? Thanks!

**Comment by blefaudeux** [comment]
- with OSS and DDP it should work I think, unless Apex is changing the parameter buffers behind the scene. Maybe worth trying if you already have one of these in place in your codebase, if not probably not a future proof enough idea
- with OSS and ShardedDDP it will probably not work, because the gradients are sharded and the Apex scaler is not aware of that, so different ranks will have different behaviors (and collective operations will eventually deadlock)

In general PytorchAMP is the supported solution moving forward (nvidia deprecated Apex some time ago as Anjali mentioned). It is a bit on the slow side though, and if you want more speed then FSDP is a better bet (it comes with its own mixed precision handling which is faster than Pytorch on distributed problems)

**Comment by amsword** [comment]
@anj-s, in some dataset/network, we find fully fp16 network works as well as AMP in terms of accuracy, but fp16 could save much more memory, in which case we can use larger batch size (normally 2x). In the end, fp16 is much faster than pytorch's AMP (1.5x~2x speed up). Thus, in some cases, we prefer to use fp16_optimizer. This is also the default setting from deepspeed (if i understand correctly). We know that in some other cases, fully fp16 leads to much worse accuracy, but not all cases.

**Comment by blefaudeux** [comment]
> @anj-s, in some dataset/network, we find fully fp16 network works as well as AMP in terms of accuracy, but fp16 could save much more memory, in which case we can use larger batch size (normally 2x). In the end, fp16 is much faster than pytorch's AMP (1.5x~2x speed up). Thus, in some cases, we prefer to use fp16_optimizer. This is also the default setting from deepspeed (if i understand correctly). We know that in some other cases, fully fp16 leads to much worse accuracy, but not all cases.

Makes sense ! Couple of points here:
- if full fp16 works for you, then you can use it directly with OSS, it should just work and be pretty fast (at least a lot faster than AMP) & save memory.
- if using AMP, you can switch on the broadcast fp16 compression in OSS, it should help a little
- Apex is faster than Pytorch AMP in a distributed setup, among other things, because the gradients are fp16 (they're promoted to fp32 with AMP), so the network communication is way smaller/faster
- For a better handling of mixed precision, I would reiterate that FSDP is the way to go, the gradients are reduced as fp16 (cc @min-xu-ai  for sanity checking) for instance, it should be very close or better than any Apex-related setup in terms of speed

**Comment by amsword** [comment]
> 
> * if full fp16 works for you, then you can use it directly with OSS, it should just work and be pretty fast (at least a lot faster than AMP) & save memory.

------
FP16_optimizer maintains the fp32 copy of the parameters and takes fp32 params as master copy. The gradient update is on these master params, after which the updated master param updates the fp16 parameters for next iteration. I'm not sure how useful this logic is in practice, but with OSS, it seems like this logic will not be handled automatically. Is my understanding correct?

**Comment by blefaudeux** [comment]
> > * if full fp16 works for you, then you can use it directly with OSS, it should just work and be pretty fast (at least a lot faster than AMP) & save memory.
> 
> FP16_optimizer maintains the fp32 copy of the parameters and takes fp32 params as master copy. The gradient update is on these master params, after which the updated master param updates the fp16 parameters for next iteration. I'm not sure how useful this logic is in practice, but with OSS, it seems like this logic will not be handled automatically. Is my understanding correct?

You mentioned "fully fp16 network" above, so I thought that you meant `model.half()`, which would work for OSS (everything is fp16).  From a quick read of the doc, I think that the Apex optimizer_fp16 should work with OSS+DDP, but not with OSS+ShardedDDP, it has never been tested though but if it's easy to try, go for it !

> Just found mixed_precision option in FSDP constructor. Setting it to True should have similar logic about fp32/fp16. Is this right?
Yes, even a little more complex than that if I remember correctly (the model update is computed in fp32 but after that the model is fp16, at least that wast the behaviour in FairSeq for memory-efficient fp16 and FSDP is related). I could be wrong on that and it's not overly important, in short yes with FSDP you get more efficient mixed precision and overall it's the more future proof solution if you're after speed and memory savings.

**Comment by amsword** [comment]
@blefaudeux  thanks. Regarding Fp16_optimizer + OSS, one thing is that FP16_opitmizer assumes init_optimizer.param_groups contains all trainable parameters and it handles to update fp32 and fp32 -> fp16. After checking the code of OSS, my understanding is that OSS.param_groups also contains all parameters but only OSS.optim.param_groups contains the sharded parameters. So, if OSS wrapps FP16_optimizer, it should not work. Is this correct? If FP16_optimizer wraps OSS, intuitively it should work as you commented, right?


I follow https://fairscale.readthedocs.io/en/latest/tutorials/oss.html, and write a small code snippet as follows, but the wrapped module is not of FSDP type. Pytorch: 1.7.1; fairscale: 0.3.7.
Is there some other example codes I can check/learn to play with FSDP?

```python
      from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP                                                                  
      fsdp_params = dict(mixed_precision=True, flatten_parameters=True)                                                                        
      from fairscale.nn.wrap import auto_wrap, enable_wrap                                                                                     
      model = torch.nn.Transformer()                                                                                                           
      with enable_wrap(wrapper_cls=FSDP, **fsdp_params):                                                                                       
          model = auto_wrap(model)                                                                                                             
      print(type(model))     
# output: class 'torch.nn.modules.transformer.Transformer'
```

**Comment by blefaudeux** [comment]
> @blefaudeux  thanks. Regarding Fp16_optimizer + OSS, one thing is that FP16_opitmizer assumes init_optimizer.param_groups contains all trainable parameters and it handles to update fp32 and fp32 -> fp16. After checking the code of OSS, my understanding is that OSS.param_groups also contains all parameters but only OSS.optim.param_groups contains the sharded parameters. So, if OSS wrapps FP16_optimizer, it should not work. Is this correct? If FP16_optimizer wraps OSS, intuitively it should work as you commented, right?

Hmm, depends on whether the parameters are rewritten in place or not, if they are I think that it would work given the parameter flattening in OSS.


> I follow https://fairscale.readthedocs.io/en/latest/tutorials/oss.html, and write a small code snippet as follows, but the wrapped module is not of FSDP type. Pytorch: 1.7.1; fairscale: 0.3.7.
> 
> Is there some other example codes I can check/learn to play with

https://fairscale.readthedocs.io/en/latest/tutorials/oss.html
Has an example at the bottom, the url is misleading. You just need FSDP in that case, no need to mix with OSS or anything


> 
> 
> ```python
> 
>       from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP                                                                  
> 
>       fsdp_params = dict(mixed_precision=True, flatten_parameters=True)                                                                        
> 
>       from fairscale.nn.wrap import auto_wrap, enable_wrap                                                                                     
> 
>       model = torch.nn.Transformer()                                                                                                           
> 
>       with enable_wrap(wrapper_cls=FSDP, **fsdp_params):                                                                                       
> 
>           model = auto_wrap(model)                                                                                                             
> 
>       print(type(model))     
> 
> # output: class 'torch.nn.modules.transformer.Transformer'
> 
> ```
> 
>  
> 
>

**Comment by min-xu-ai** [comment]
Hi @blefaudeux , this could be closed now?

**Comment by blefaudeux** [comment]
> Hi @blefaudeux , this could be closed now?

I think so ? in essence this feels close to "memory efficient fp16" that FSDP does, if this works out of the box with OSS+Apex that's nice, but if not I think that FSDP is the reasonable path forward to get this feature set. cc @amsword if you would like something done and I misunderstood ?

**Comment by amsword** [comment]
tried to use OSS as the optimizer for apex initialization. It is runnable, but the loss is much worse. In the following, the orange curve is the one with OSS + apex.initialize (O2), while the other two is apex.initialize O2 or pytorch/amp+OSS. 

![image](https://user-images.githubusercontent.com/3356680/132938717-46b70dbe-4b64-424c-bd63-4120a876c2a5.png)

**Comment by blefaudeux** [comment]
oh thanks for testing ! There's probably something wrong with the scaling then, could for instance be that not all parameters are scaled in the same way, which could make sense since the Apex optimizer is not shard aware. Good to know !

**Comment by amsword** [comment]
finally, I figured it out, and the code is here
https://github.com/amsword/scaleapex
